rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,664 @@
+/*
+Copyright (c) 2025 Advanced Micro Devices, Inc. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <numeric>
+#include <random>
+#include <math.h>
+#include "pipeline/commons.h"
+#include "pipeline/filesystem.h"
+#include "readers/image/numpy_data_reader.h"
+
+#define CALL_AND_CHECK_FLAG(func) \
+    do { \
+        func; \
+        if (_header_parsing_failed) { \
+            goto error; \
```

Comment:
we shouldn't use goto statements in the code as it is not recommended in any our libraries. It can produce unwanted outcomes. Please rewrite avoiding goto